### DISCUSSION TITLE
[Question] Performance: manifold3d python

### DISCUSSION THREAD

**Discussion by AlbertoVPersonal** [elalish/manifold#1118] [Q&A]
Hi,

I would like to know whether the team has some reports with performance results using `mainfold3d` Python package.
We are using the JS library and the performance is not good as we need.
And we are thinking to move the code to Python or Rust.

Regards,
Alberto

**Comment by pca006132** [comment]
A native build would be faster, and it can use multiple threads. For general performance, it depends on how you use the library, we cannot comment on this without looking at your use case, and how fast you need it to be.

**Reply by AlbertoVPersonal** [reply]
Hi,

I have been talking with a colleague and he has told me this. I hope this can help you understand our issue.

> We need to apply several operations to one threejs generated mesh.
>
> First we calculate the operations by subtracting material to the operation itself if it has "holes" inside. Then we subtract the "operation" to the actual mesh. Finally we get the mesh back to the threejs ecosystem.
>
> We apply this process several times, and the perfomance downgrades as the number of operations increases. We understand that this issue could be predicted, the resulting mesh increases in complexity with the previous operations applied.
>
> Now we are trying to improve the performance as much as possible and working with the javascript binding seemed pretty straightforward at first but as we cannot benefit from the multithreading feature and we see the migration step as near as ever.

**Reply by pca006132** [reply]
We optimize on several consecutive mesh boolean operations, but to benefit from that you have to avoid frequent conversion to/from meshes. See https://github.com/elalish/manifold/wiki/Performance-Considerations#users-related

And what level of performance do you want? How much faster than wasm? Are you doing many small operations (e.g. expecting 10k operations in 1s) or just a few larger operations (10 slow operations in 1s)? Native will speed up those larger operations due to parallelization, I would not expect small operations to benefit significantly.

**Reply by AlbertoVPersonal** [reply]
I think that my colleague will reply with the details. I will talk with him tomorrow.

**Reply by elalish** [reply]
Also #1079 hasn't made it to an npm release yet, but that parallelism in WASM could have a significant speedup as well. You may want to try building that locally - I'd love to know what effect is has on performance for your use case.

**Reply by AlbertoVPersonal** [reply]
Ok. I will talk with him about your comment. Thanks!

**Reply by pca006132** [reply]
Note that #1079 is not really stable, we still encounter weird stack issue that is impossible to debug due to lack of sanitizer support in emscripten. And regular builds don't encounter any stack issues...
